Source/WebCore/platform/network/soup/NetworkStorageSessionSoup.cpp

```diff
@@ -531,136 +622,248 @@ Vector<Cookie> NetworkStorageSession::getCookies(const URL& url)
 #else
-    GUniquePtr<GSList> cookies(soup_cookie_jar_get_cookie_list(cookieStorage(), uri.get(), TRUE));
-    UNUSED_PARAM(firstParty);
-    UNUSED_PARAM(sameSiteInfo);
-    UNUSED_PARAM(firstParty);
+    CookieList cookies(soup_cookie_jar_get_cookie_list(cookieStorage(), uri.get(), forHTTPHeader == ForHTTPHeader::Yes));
 #endif
     if (!cookies)
-        return false;
-
-    for (GSList* iter = cookies.get(); iter; iter = g_slist_next(iter)) {
-        SoupCookie* soupCookie = static_cast<SoupCookie*>(iter->data);
-        rawCookies.append(Cookie(soupCookie));
-        soup_cookie_free(soupCookie);
-    }
-
-    return true;
-}
+        return nullptr;
 
-static std::pair<String, bool> cookiesForSession(const NetworkStorageSession& session, const URL& firstParty, const URL& url, const SameSiteInfo& sameSiteInfo, std::optional<FrameIdentifier> frameID, std::optional<PageIdentifier> pageID, bool forHTTPHeader, IncludeSecureCookies includeSecureCookies, ApplyTrackingPrevention applyTrackingPrevention, ShouldRelaxThirdPartyCookieBlocking relaxThirdPartyCookieBlocking)
-{
-    if (applyTrackingPrevention == ApplyTrackingPrevention::Yes && session.shouldBlockCookies(firstParty, url, frameID, pageID, relaxThirdPartyCookieBlocking))
-        return { { }, false };
-
-    auto uri = urlToSoupURI(url);
-    if (!uri)
-        return { { }, false };
-
-#if SOUP_CHECK_VERSION(2, 69, 90)
-    auto firstPartyURI = urlToSoupURI(firstParty);
-    if (!firstPartyURI)
-        return { { }, false };
-
-    auto cookieURI = sameSiteInfo.isSameSite ? urlToSoupURI(url) : nullptr;
-    GSList* cookies = soup_cookie_jar_get_cookie_list_with_same_site_info(session.cookieStorage(), uri.get(), firstPartyURI.get(), cookieURI.get(), forHTTPHeader, sameSiteInfo.isSafeHTTPMethod, sameSiteInfo.isTopSite);
-#else
-    GSList* cookies = soup_cookie_jar_get_cookie_list(session.cookieStorage(), uri.get(), forHTTPHeader);
-    UNUSED_PARAM(firstParty);
-    UNUSED_PARAM(sameSiteInfo);
-#endif
-    bool didAccessSecureCookies = false;
+    bool accessedSecureCookies = false;
 
     // libsoup should omit secure cookies itself if the protocol is not https.
     if (url.protocolIs("https"_s)) {
-        GSList* item = cookies;
-        while (item) {
+        for (GSList* item = cookies.get(); item; item = g_slist_next(item)) {
             auto cookie = static_cast<SoupCookie*>(item->data);
             if (soup_cookie_get_secure(cookie)) {
-                didAccessSecureCookies = true;
+                accessedSecureCookies = true;
                 if (includeSecureCookies == IncludeSecureCookies::No) {
-                    GSList* next = item->next;
                     soup_cookie_free(static_cast<SoupCookie*>(item->data));
-                    cookies = g_slist_remove_link(cookies, item);
-                    item = next;
+                    auto* oldList = cookies.release();
+                    cookies = CookieList(g_slist_remove_link(oldList, item));
```

Comment:
Is `item` leaked?

Reply:
Yeah, always was. I assumed `g_slist_remove_link` was `g_slist_delete_link`.